### PR TITLE
Fabfile changes

### PR DESCRIPTION
**Pull request by maparent**

### CODE REVIEW

File: fabfile.py [hunk 1]
```diff
@@ -794,39 +828,92 @@ def _install_builddeps():
...
-        sudo('apt-get install -y graphviz libgraphviz-dev pkg-config')
-        sudo('apt-get install -y liblapack-dev libatlas-dev libblas-dev gfortran')
-        print "We are still trying to get some requirements right for linux,"\
-            "See http://www.scipy.org/scipylib/building/linux.html for details."
-
-        # Runtime requirements (even in develop)
-        sudo('apt-get install -y redis-server memcached unixodbc-dev')
-    execute(install_database)
-    execute(update_python_package_builddeps)
+        sudo('apt-get install -y automake bison flex gperf gawk')
```
Comment (ayyazdaniaryan) [inline fabfile.py]: Where did these dependencies originate from @maparent ? automake, bison, flex, gfortran, etc...

File: fabfile.py [hunk 2]
```diff
@@ -794,39 +828,92 @@ def _install_builddeps():
...
+             'libxslt1-dev libffi-dev libhiredis-dev libxml2-dev libssl-dev '
+             'libreadline-dev liblapack-dev libatlas-dev libblas-dev '
+             'libgraphviz-dev')
+        print ("We are still trying to get some requirements right for linux, "
+               "See http://www.scipy.org/scipylib/building/linux.html "
+               "for details.")
+
+
+@task
+def install_redis():
```
Comment (ayyazdaniaryan) [inline fabfile.py]: Me gusta mucho! This separation is very helpful in future automation!

File: fabfile.py [hunk 3]
```diff
@@ -1820,3 +1955,40 @@ def env_bel_bluenove():
...
+            continue
+        fname = join('configs', service, config_name + ".rc")
+        if os.path.exists(fname):
+            service_config = load_settings(fname)
+            for k, v in service_config.iteritems():
+                k = "_".join((service, k))
+                if k not in env:
+                    env[k] = v
+
+load_service_configs()
```
Comment (ayyazdaniaryan) [inline fabfile.py]: Would putting the function here mean that as soon as the fabfile is loaded by the fabric deamon, the method is run?

File: fabfile.py [hunk 4]
```diff
@@ -1820,3 +1955,40 @@ def env_bel_bluenove():
...
+    for service in SERVICES:
+        config_name = env.get(service + "_config", "default")
+        if config_name is None:
+            continue
+        fname = join('configs', service, config_name + ".rc")
+        if os.path.exists(fname):
+            service_config = load_settings(fname)
+            for k, v in service_config.iteritems():
+                k = "_".join((service, k))
+                if k not in env:
```
Comment (ayyazdaniaryan) [inline fabfile.py]: re-using `k` makes this hard to read.

File: fabfile.py [hunk 5]
```diff
@@ -1820,3 +1955,40 @@ def env_bel_bluenove():
...
+    'piwik',
+    'sentry',
+    'memcached',
+    'redis',
+    'mysql',
+]
+
+
+def load_service_configs():
+    """ allows to break a service network configuration into many files.
```
Comment (ayyazdaniaryan) [inline fabfile.py]: The documentation is not enough. Let's give more:

```
eg. Assume that there is a server environment, called X.

X.rc samples like the following:

postgres_config = {
  'db_user': 'assembl',
  'db_password': 'assembl' 
}

This configuration, whether it exists or not in the default ENV variable, will now exist, as a result of this method.

File: fabfile.py [hunk 6]
```diff
@@ -1508,6 +1641,8 @@ def commonenv(projectpath, venvpath=None):
     # It is recommended you keep localhost even if you have access to
     # unix domain sockets, it's more portable across different pg_hba configurations.
     env.db_host = 'localhost'
+    if config.has_option("app:assembl", "db_host"):
```
Comment (ayyazdaniaryan) [inline fabfile.py]: The host is important for pg, for what reason?

File: fabfile.py [hunk 7]
```diff
@@ -1316,6 +1408,47 @@ def install_database():
...
+    Installs Apache2, Mysql and PHP on a Linux Environment
+    """
+    if env.mac:
+        print(red("This task cannot be run on a Macintosh, you fool!"))
+    else:
+        sudo("apt-get purge apache2 mysql-server php-mysql php-curl php-cli php-gd")
+        sudo("apt-get autoremove")  # Remove dangling dependencies after purging
+
+@task
+def install_piwik():
```
Comment (ayyazdaniaryan) [inline fabfile.py]: This task is not yet tested. By me!

File: fabfile.py [hunk 8]
```diff
@@ -794,39 +828,92 @@ def _install_builddeps():
...
+        run('brew install memcached')
+        run('brew tap homebrew/services')
+        run('brew services start memcached')
     else:
-        sudo('apt-get install -y postgresql postgresql-server-dev-all libmemcached-dev libzmq3-dev libxslt1-dev libffi-dev phantomjs')
+        sudo('apt-get install -y memcached')
+
+
+@task
+def set_file_permissions():
```
Comment (ayyazdaniaryan) [inline fabfile.py]: I appreciate this, as setting permissions -> not my most high quality skill.

File: fabfile.py [hunk 9]
```diff
@@ -756,19 +732,77 @@ def update_npm_requirements(force_reinstall=False):
...
+    execute(install_database)
+    execute(install_assembl_server_deps)
+
+
+@task
+def install_assembl_server_deps():
+    """
+    Will install most assembl components on a single server, except db
+    """
+    sanitize_env()
```
Comment (ayyazdaniaryan) [inline fabfile.py]: Does `sanitize_env` behave exactly as it should if the env is loaded by an `.rc` file?

File: fabfile.py (hunk 1 above)
Comment (ayyazdaniaryan) [inline fabfile.py]: Aka...I understand they are needed to proceed with Assembl install, no clue as to how they were identified.

File: fabfile.py [hunk 10]
```diff
@@ -794,39 +828,92 @@ def _install_builddeps():
...
+        run('find assembl/static -type f -print|xargs chmod g+r')
+        # allow postgres user to use pypsql
+        run('chmod go+x . assembl assembl/scripts')
+        run('chmod go+r assembl/scripts/pypsql.py')
+    with cd(env.venvpath):
+        run('chmod go+rx . bin')
+        run('chmod -R go+r lib')
+        run('find lib -type d -print|xargs chmod go+x')
+        run('chmod go+x bin/python')
+        run('chmod go+r bin/activate')
```
Comment (ayyazdaniaryan) [inline fabfile.py]: go+r -> can be executed?

File: fabfile.py [hunk 11]
```diff
@@ -1820,3 +1955,40 @@ def env_bel_bluenove():
     env.uses_ngnix = True
     env.uses_uwsgi = True
     env.gitbranch = getenv("GITBRANCH", "master")
+
+
+""" The various services used by the stack """
```
Comment (ayyazdaniaryan) [inline fabfile.py]: Should we move this to the very top of the file for ease of readability/and immediate invocation of the method below?

File: fabfile.py (hunk 9 above)
Comment (maparent) [inline fabfile.py]: load_environment gives us strings; env variables declared in the fabfile env_... can be strings, lists or bools. sanitize_env ensures type uniformity between both origins.

File: fabfile.py (hunk 10 above)
Comment (maparent) [inline fabfile.py]: read by group and others.

File: fabfile.py (hunk 6 above)
Comment (maparent) [inline fabfile.py]: very different behaviour if talking to postgres through a tcp socket (other host) or pg_ctl (local)

File: fabfile.py (hunk 3 above)
Comment (maparent) [inline fabfile.py]: yes.

File: fabfile.py (hunk 4 above)
Comment (maparent) [inline fabfile.py]: If you think so.

File: fabfile.py (hunk 11 above)
Comment (maparent) [inline fabfile.py]: Moving it up is possible.

File: fabfile.py (hunk 5 above)
Comment (maparent) [inline fabfile.py]: Not accurate as given. I'll give it another stab.

File: fabfile.py (hunk 7 above)
Comment (maparent) [inline fabfile.py]: Well we're merging wip here.

File: fabfile.py [hunk 12]
```diff
@@ -900,14 +1030,16 @@ def check_and_create_database_user():
     with settings(warn_only=True):
         checkUser = venvcmd('assembl-pypsql -1 -u {user} -p {password} -n {host} "{command}"'.format(
             command="SELECT 1 FROM pg_roles WHERE rolname='%s'" % (env.db_user),
-            python=env.venvpath + "/bin/python", password=env.db_password,
+            python=join(env.venvpath, "bin", "python"), password=env.db_password,
             host=env.db_host, user=env.db_user, projectpath=env.projectpath))
     if checkUser.failed:
         print(yellow("User does not exist"))
-        run_db_command(as_venvcmd('assembl-pypsql -u {db_user} "{command}"'.format(
+        run_db_command(as_venvcmd('{python} {pypsql} -u {db_user} -n {host} "{command}"'.format(
```
Comment (swergas) [inline fabfile.py]: This still does not work: postgres user is not created and console output is the following.

```
User does not exist
[localhost] sudo: . /home/assembl_user/assembl/venv/bin/activate && /home/assembl_user/assembl/venv/bin/python /home/assembl_user/assembl/assembl/scripts/pypsql.py -u postgres -n localhost "CREATE USER assembl WITH CREATEDB ENCRYPTED PASSWORD 'assembl'; COMMIT;"
[localhost] out: /bin/bash: /root/.bash_profile: Permission denied
[localhost] out: /bin/bash: /home/assembl_user/assembl/venv/bin/activate: Permission denied
[localhost] out: 


Fatal error: sudo() received nonzero return code 1 while executing!

Requested: . /home/assembl_user/assembl/venv/bin/activate && /home/assembl_user/assembl/venv/bin/python /home/assembl_user/assembl/assembl/scripts/pypsql.py -u postgres -n localhost "CREATE USER assembl WITH CREATEDB ENCRYPTED PASSWORD 'assembl'; COMMIT;"
Executed: sudo -S -p 'sudo password:'  -u "postgres"  /bin/bash -l -c ". /home/assembl_user/assembl/venv/bin/activate && /home/assembl_user/assembl/venv/bin/python /home/assembl_user/assembl/assembl/scripts/pypsql.py -u postgres -n localhost \"CREATE USER assembl WITH CREATEDB ENCRYPTED PASSWORD 'assembl'; COMMIT;\""

Aborting.
Disconnecting from localhost... done.
```

File: fabfile.py (hunk 6 above)
Comment (ayyazdaniaryan) [inline fabfile.py]: Cool, so this enables automation of remote postgres server?

Comment (maparent) [inline fabfile.py]: correct.

**Pull request merged**